File: frontend/src/features/dashboards/spec.ts
// The WIDGET SPEC, browser side — spec_version 2.
//
// This is the same object `backend/reading-writer/app/api/spec.py` validates and
// `sqlgen.py` turns into SQL. Read those for the reasoning; this file is the
// TypeScript view plus the small amount of presentation knowledge the backend
// deliberately does not have.
//
// WHAT CHANGED FROM v1 AND WHY
// ----------------------------
// v1 had an IoT vocabulary baked into it — `scope: points | device | category |
// all`, `metric: avg | … | count`. It could not chart a door-access event or a
// fire-panel state, so it could not be the platform's one builder. v2 names a
// DATASET from the registry and picks dimensions, measures and aggregates out of
// whatever that dataset publishes. Nothing in this file knows what a "point" is.
//
// The split to keep in mind while editing:
//
//   query.*    is CONTRACT. It goes to the backend, which validates it with
//              `extra="forbid"`. Adding a field here without adding it there is
//              a 400 naming the field, not a silent no-op.
//   viz        is ADVISORY. The backend never validates it; every result is the
//              same `{columns, rows}` table, so a new chart needs no backend
//              change at all.
//   options.*  is OURS ALONE. Presentation only, stored verbatim, never parsed
//              by the backend.
//
// **There is no `sql` field, here or anywhere.** The client sends STATE; the
// server generates the statement. See the backend spec module — that decision is
// not reopened.

export const SPEC_VERSION = 2;

export type Viz = "line" | "bar" | "stat" | "table";

/** Aggregates the builder can ask for. Ported from the reference product's
 *  `AGGREGATE_OPTIONS`, plus first/last which this platform's stores can answer.
 *  Which of them a given measure PERMITS comes from the dataset, not from here. */
export type Aggregate =
  | "count"
  | "count_distinct"
  | "sum"
  | "avg"
  | "min"
  | "max"
  | "first"
  | "last";

/** Filter operators. Ported verbatim from their `FILTER_OP_OPTIONS`. */
export type FilterOp =
  | "="
  | "!="
  | "<"
  | "<="
  | ">"
  | ">="
  | "contains"
  | "like"
  | "in"
  | "between"
  | "is null"
  | "is not null";

export const AGGREGATE_LABEL: Record<Aggregate, string> = {
  count: "Count",
  count_distinct: "Count distinct",
  sum: "Sum",
  avg: "Average",
  min: "Min",
  max: "Max",
  first: "First",
  last: "Last",
};

export const FILTER_OP_LABEL: Record<FilterOp, string> = {
  "=": "equals",
  "!=": "not equals",
  ">": "greater than",
  ">=": "greater or equal",
  "<": "less than",
  "<=": "less or equal",
  contains: "contains",
  like: "like (pattern)",
  in: "is one of",
  between: "between",
  "is null": "is empty",
  "is not null": "is not empty",
};

/** Operators that take no value at all — the form hides the value box for these. */
export const NO_VALUE_OPS: FilterOp[] = ["is null", "is not null"];

// ── the dataset registry, as the browser sees it ────────────────────────────

export interface DatasetDimension {
  key: string;
  label: string;
  type: "text" | "uuid" | "number" | "bool" | "time";
  description?: string;
}

export interface DatasetMeasure {
  key: string;
  label: string;
  type: string;
  /** Which aggregates this measure permits. A measure that must not be summed
   *  says so HERE, and the editor offers only what is listed. */
  aggregates: Aggregate[];
  description?: string;
  /** False when values are not comparable across series — the honesty rule
   *  (contract §4). The editor steers instead of letting the user discover it as
   *  a 422. */
  comparable: boolean;
  comparable_within: string[];
  incomparable_hint?: string;
}

export interface DatasetResolution {
  key: string;
  grain_sec: number;
  max_window_minutes: number | null;
  reason: string;
}

export interface Dataset {
  key: string;
  name: string;
  description: string;
  permission: string;
  dimensions: DatasetDimension[];
  measures: DatasetMeasure[];
  resolutions: DatasetResolution[];
  defaults: {
    series_by?: string | null;
    label_dimension?: string | null;
    measure?: string | null;
    aggregate?: Aggregate | null;
  };
}

// ── builder state ───────────────────────────────────────────────────────────

export interface SelectItem {
  dimension?: string | null;
  measure?: string | null;
  aggregate?: Aggregate | null;
  alias?: string | null;
}

export interface Filter {
  column: string;
  op: FilterOp;
  value?: string | number | null;
  value2?: string | number | null;
  values?: (string | number)[];
}

export interface Having {
  measure: string;
  aggregate: Aggregate;
  op: FilterOp;
  value?: number | null;
  value2?: number | null;
}

export interface OrderBy {
  select_index: number;
  dir: "asc" | "desc";
}

export interface SpecWindow {
  last_hours?: number | null;
  start?: string | null;
  end?: string | null;
}

export interface SpecQuery {
  dataset: string;
  resolution: string;
  window: SpecWindow;
  select: SelectItem[];
  time_series?: boolean;
  series_by?: string | null;
  series_label?: string | null;
  filters?: Filter[];
  filter_combinator?: "AND" | "OR";
  group_by?: string[];
  having?: Having[];
  order_by?: OrderBy[];
  limit: number;
  band?: boolean;
}

export interface WidgetSpec {
  spec_version: number;
  viz: string;
  query: SpecQuery;
  options?: Record<string, any>;
}

// ── the result ──────────────────────────────────────────────────────────────
//
// ONE shape, for every dataset and every chart: named columns and positional
// rows. It is the reference product's chart-data contract, which is what lets a
// chart type be added without touching the query layer.

export interface QueryResult {
  shape: "table";
  dataset: string;
  columns: string[];
  rows: (string | number | null)[][];
  label_index: number;
  resolution: string;
  resolution_reason: string;
  start: string;
  end: string;
  matched: number;
  truncated: boolean;
  band: [number | null, number | null][] | null;
  /** A read-only ECHO of the statement the SERVER generated. Shown in the
   *  builder so a person can see exactly what will run. Nothing anywhere accepts
   *  SQL back — see the backend spec module. */
  sql: string;
}

// ── steering ────────────────────────────────────────────────────────────────

/** Which chart types are time-series shaped. A line chart asks for buckets over
 *  time; a bar/stat/table asks for one row per group. Flipping this when the
 *  chart type changes is what stops a bar chart asking for time buckets it
 *  cannot draw. */
export const VIZ_TIME_SERIES: Record<Viz, boolean> = {
  line: true,
  bar: false,
  stat: false,
  table: false,
};

export const WINDOWS: { hours: number; label: string }[] = [
  { hours: 1, label: "1H" },
  { hours: 6, label: "6H" },
  { hours: 24, label: "24H" },
  { hours: 24 * 7, label: "7D" },
  { hours: 24 * 30, label: "30D" },
];

export const measureOf = (ds: Dataset | undefined, key?: string | null) =>
  ds?.measures.find((m) => m.key === key);
export const dimensionOf = (ds: Dataset | undefined, key?: string | null) =>
  ds?.dimensions.find((d) => d.key === key);

/** A blank widget over a dataset, using the dataset's OWN declared defaults —
 *  so a domain that registers tomorrow gets a sensible starting widget without
 *  anything here knowing its column names. */
export function newSpec(viz: Viz, ds?: Dataset): WidgetSpec {
  const timeSeries = VIZ_TIME_SERIES[viz];
  const measure = ds?.defaults.measure || ds?.measures[0]?.key || "";
  const m = measureOf(ds, measure);
  const aggregate = (ds?.defaults.aggregate && m?.aggregates.includes(ds.defaults.aggregate)
    ? ds.defaults.aggregate
    : m?.aggregates[0]) as Aggregate | undefined;
  const seriesBy = ds?.defaults.series_by || null;
  const label = ds?.defaults.label_dimension || null;

  const select: SelectItem[] = timeSeries
    ? [{ measure, aggregate }]
    : [
        ...(label ? [{ dimension: label } as SelectItem] : []),
        { measure, aggregate },
      ];

  return {
    spec_version: SPEC_VERSION,
    viz,
    query: {
      dataset: ds?.key || "",
      resolution: "auto",
      window: { last_hours: 6 },
      select,
      time_series: timeSeries,
      series_by: timeSeries ? seriesBy : null,
      series_label: timeSeries && seriesBy && label && label !== seriesBy ? label : null,
      filters: [],
      filter_combinator: "AND",
      group_by: timeSeries ? [] : label ? [label] : [],
      having: [],
      order_by: timeSeries ? [] : [{ select_index: select.length - 1, dir: "desc" }],
      limit: viz === "stat" ? 1 : viz === "line" ? 4 : 8,
    },
    options: {},
  };
}

/** Is this spec answerable? Mirrors the backend's rules so the editor can say
 *  what is missing instead of firing a request that comes back 422.
 *
 *  The comparability check is the one that matters: it is contract §4's "a value
 *  metric cannot be grouped across incomparable series", enforced on the server
 *  and MIRRORED here so the editor steers rather than traps. */
export function specIssue(spec: WidgetSpec, ds?: Dataset): string | null {
  const q = spec.query;
  if (!q.dataset) return "Pick a dataset.";
  if (!q.select.length) return "Add at least one column or measure.";
  if (!ds) return null;

  for (const item of q.select) {
    if (item.measure) {
      const m = measureOf(ds, item.measure);
      if (!m) return `This dataset has no measure “${item.measure}”.`;
      if (!item.aggregate || !m.aggregates.includes(item.aggregate)) {
        return `Pick an aggregate for “${m.label}”.`;
      }
    }
  }
  if (q.time_series && q.series_by && q.select.filter((s) => s.measure).length !== 1) {
    return "A split time-series draws exactly one measure.";
  }

  const pinned = new Set<string>([...(q.group_by || [])]);
  if (q.series_by) pinned.add(q.series_by);
  for (const f of q.filters || []) {
    if (f.op === "=" && f.value !== undefined && f.value !== null && `${f.value}` !== "") {
      pinned.add(f.column);
    }
    if (f.op === "in" && (f.values || []).length === 1) pinned.add(f.column);
  }
  for (const item of q.select) {
    if (!item.measure) continue;
    if (item.aggregate === "count" || item.aggregate === "count_distinct") continue;
    const m = measureOf(ds, item.measure);
    if (!m || m.comparable) continue;
    if (m.comparable_within.some((k) => pinned.has(k))) continue;
    const names = m.comparable_within
      .map((k) => dimensionOf(ds, k)?.label || k)
      .join(", ");
    return `“${AGGREGATE_LABEL[item.aggregate!]} of ${m.label}” cannot be computed across mixed series: ${
      m.incomparable_hint || "values from different series are not comparable."
    } Group by (or filter to) one of: ${names}.`;
  }
  return null;
}

/** Format a number for display. NO unit is ever appended — a dataset can name a
 *  unit column but nothing here invents one, and a fabricated "kW" on an axis is
 *  worse than a blank (contract §4). */
export function fmtValue(v: number | null | undefined, decimals?: number): string {
  if (v === null || v === undefined || !Number.isFinite(v)) return "—";
  if (typeof decimals === "number") return v.toFixed(decimals);
  const abs = Math.abs(v);
  if (abs === 0) return "0";
  // An exact integer prints as one. Without this a sample COUNT of 12 renders as
  // "12.0", which reads as a measurement rather than a tally.
  if (Number.isInteger(v)) return v.toLocaleString();
  if (abs >= 1000) return v.toLocaleString(undefined, { maximumFractionDigits: 0 });
  if (abs >= 10) return v.toFixed(1);
  if (abs >= 1) return v.toFixed(2);
  if (abs < 0.001) return v.toExponential(2);
  return v.toFixed(3);
}

// ── v1 → v2 migration (contract §6) ─────────────────────────────────────────
//
// Saved widgets exist against the v1, IoT-shaped spec (`scope` / `metric` /
// `group_by`). The backend migrates one on every read; this is the SAME
// translation on the browser side, and it is needed for a reason worth stating:
// the editor and the client-side validator both work against v2 state, so a v1
// spec reaching them unmigrated would show "Pick a dataset" over a widget that
// the server can answer perfectly well. Keeping the two in step is what makes a
// saved dashboard keep rendering rather than going blank.
//
// It is a mirror of `backend/reading-writer/app/api/builder.py::migrate_v1`. If
// one changes the other must; that is the cost of having a client-side validator
// at all, and it is cheaper than a dashboard that only reports its errors from
// the server.

const V1_METRIC: Record<string, [string, Aggregate]> = {
  avg: ["value", "avg"],
  min: ["value", "min"],
  max: ["value", "max"],
  first: ["value", "first"],
  last: ["value", "last"],
  // v1's `count` was SAMPLES — a tally of readings, not a physical quantity.
  count: ["samples", "sum"],
};

const V1_GROUP_DIM: Record<string, string> = {
  point: "point_id",
  device: "device_tag",
  category: "category",
};

function migrateV1(raw: any): WidgetSpec {
  const q = raw?.query || {};
  const scope = q.scope || {};
  const metric: string = q.metric || "avg";
  const [measure, aggregate] = V1_METRIC[metric] || ["value", "avg"];
  const filters: Filter[] = [];

  const stype = scope.type || "points";
  if (stype === "points") {
    filters.push({ column: "point_id", op: "in", values: (scope.point_ids || []).map(String) });
  } else if (stype === "device") {
    if (scope.device_id) filters.push({ column: "device_id", op: "=", value: String(scope.device_id) });
    else if (scope.device_tag) filters.push({ column: "device_tag", op: "=", value: scope.device_tag });
  } else if (stype === "category") {
    // v1: category "" meant the points nothing has classified.
    if (scope.category === "" || scope.category == null) {
      filters.push({ column: "category", op: "is null" });
    } else {
      filters.push({ column: "category", op: "=", value: scope.category });
    }
  }
  if (metric !== "count") {
    // v1 restricted value metrics to numeric points: a text point has no number,
    // and including it would show a permanently blank row.
    filters.push({ column: "reading_kind", op: "=", value: "num" });
  }

  const window = q.window || { last_hours: 6 };
  const limit = Number(q.limit || 12) || 12;
  const viz = raw?.viz || "line";

  let query: SpecQuery;
  if ((q.kind || "series") === "series") {
    query = {
      dataset: "iot_readings",
      resolution: q.rollup || "auto",
      window,
      time_series: true,
      series_by: "point_id",
      series_label: "point_tag",
      select: [{ measure, aggregate }],
      filters,
      limit: Math.min(limit, 24),
      band: !!raw?.options?.band,
    };
  } else {
    const dim = V1_GROUP_DIM[q.group_by || "point"] || "point_id";
    const select: SelectItem[] =
      dim === "point_id"
        ? [
            { dimension: "point_tag", alias: "point" },
            { dimension: "device_tag", alias: "device" },
            { measure, aggregate, alias: metric },
            { measure: "samples", aggregate: "sum", alias: "samples" },
          ]
        : [
            { dimension: dim, alias: q.group_by },
            { measure: "samples", aggregate: "sum", alias: "samples" },
          ];
    query = {
      dataset: "iot_readings",
      resolution: q.rollup || "auto",
      window,
      time_series: false,
      select,
      group_by: dim === "point_id" ? ["point_id", "point_tag", "device_tag"] : [dim],
      order_by: [{ select_index: select.length - 1, dir: "desc" }],
      filters,
      limit,
    };
  }

  return { spec_version: SPEC_VERSION, viz, query, options: { ...(raw?.options || {}) } };
}

/** Bring a stored spec up to `SPEC_VERSION`. The ONE place that happens on the
 *  browser side. A spec from the FUTURE is left alone — the server refuses it
 *  loudly, which is better than this build guessing at fields it has never
 *  heard of. */
export function migrateSpec(raw: WidgetSpec | any): WidgetSpec {
  if (!raw || typeof raw !== "object") return raw;
  const v = typeof raw.spec_version === "number" ? raw.spec_version : SPEC_VERSION;
  if (v <= 1) return migrateV1(raw);
  return raw as WidgetSpec;
}

File: backend/reading-writer/app/api/spec.py
"""The WIDGET QUERY SPEC — a structured, versioned description of what a
dashboard widget shows. This is the contract the dashboard builder writes and
this service executes.

## Why a spec and not SQL

The obvious alternative is letting a widget carry a SQL string. It is rejected
here, deliberately and permanently:

* This platform runs video surveillance and access control. Arbitrary SQL
  arriving from a browser is a far worse trade here than in a standalone BI tool,
  because the blast radius of a guard bug is the whole console.
* Free-SQL guards are genuinely hard. The reference implementation reviewed
  before this module was written had three real defects — a leading comment was
  rejected outright, a trailing comment broke the wrapper the guard appended, and
  validation truncated at the first `;` even when that `;` was inside a string
  literal or a comment. Each is a small mistake; together they are a bypass.
* This module does not need one. The data is KNOWN: a `points` dimension and two
  continuous aggregates. Picking a scope, a metric, a window and a rollup covers
  every question the store can honestly answer — and picking rather than typing
  is what makes the builder no-code in the first place.

## Forward compatibility — the part that must not be got wrong

A saved dashboard is data with a long life. It has to survive this module
gaining widget types it has never heard of. Three rules make that true:

1. **`viz` is not validated here.** The executor does not care how a result is
   drawn; it only cares about `query.kind`. Adding a "gauge" or "heatmap" that
   reuses the `aggregate` result shape therefore needs NO backend change at all,
   and an old backend serving a new frontend keeps working.
2. **`query.kind` is the small, closed set** (`series`, `aggregate`) — the RESULT
   SHAPES, not the chart types. Four visualisations ship in v1 and they use two
   shapes between them. New shapes are a spec-version bump; new charts are not.
3. **`spec_version` is checked, and a spec from the FUTURE is refused loudly**
   rather than being half-executed with fields this build silently ignored. A
   spec from the PAST is migrated by `_migrate` (a no-op today, and the hook that
   means v1 widgets keep rendering after v2 lands).

`options` is free-form on purpose: it is presentation only (title, accent,
decimals) and never reaches the database. `extra="forbid"` on everything that DOES
reach the database is the other half of that trade — a typo'd `metrc` is a 400
naming the field, not a silently-ignored key that makes the widget show the
wrong number.

## The honesty rules this spec enforces

They are the same rules the rest of Building Intelligence lives by, moved into
the one place a widget can be built:

* **Charts read the ROLLUPS.** `rollup="raw"` is accepted only inside
  `RAW_MAX_MINUTES`; past it the widget is refused with the rollup to ask for
  instead. Nothing is silently downgraded, because a chart that quietly changes
  resolution is a chart that lies about its own precision.
* **No unit is invented.** `unit` is passed through from `points.unit` and is
  NULL for every point on this deployment, which is correct — the source payloads
  carry none.
* **A value metric cannot be grouped across points.** Averaging a power factor
  with a voltage because they hang off the same device produces a number that
  means nothing. So `group_by` of `device` or `category` is allowed ONLY for
  `count` (samples), which is a quantity the pipeline genuinely knows — the same
  reasoning `/bi/activity` already uses. Everything else groups by point.
* **Value metrics select numeric points only.** A text point ("mode", "status")
  has no `num`, and including it would show a row that is permanently blank.
"""

from __future__ import annotations

import datetime as dt
import uuid
from typing import Literal

from kernel.errors import ValidationError
from pydantic import BaseModel, ConfigDict, Field
from pydantic import ValidationError as PydanticValidationError

from . import queries as q

# The spec version this build writes and understands. A stored spec carrying a
# LOWER version is migrated on read (`_migrate`); a HIGHER one is refused.
SPEC_VERSION = 1

# How many series one widget may draw / how many rows one aggregate may return.
# A widget asking for 500 series is a bug and should fail at the edge rather than
# become a slow query. Series is the tighter cap: it is per-bucket data.
MAX_SERIES = q.MAX_SERIES_POINTS  # 24
MAX_ROWS = 100

# Widest window a widget may ask for, matching the /bi/series ceiling.
MAX_HOURS = 24 * 90


class Window(BaseModel):
    """The time range, expressed RELATIVELY by default.

    `last_hours` is the normal form and the reason a saved dashboard stays useful:
    it means "the last six hours", resolved at render time, not a window frozen on
    the afternoon somebody built the widget. Absolute `start`/`end` exist for a
    widget deliberately pinned to an incident.
    """

    model_config = ConfigDict(extra="forbid")

    last_hours: int | None = Field(default=6, ge=1, le=MAX_HOURS)
    start: dt.datetime | None = None
    end: dt.datetime | None = None

    def resolve(self) -> tuple[dt.datetime, dt.datetime]:
        now = dt.datetime.now(dt.timezone.utc)
        if self.start is not None or self.end is not None:
            end = self.end or now
            start = self.start or (end - dt.timedelta(hours=self.last_hours or 6))
        else:
            end = now
            start = end - dt.timedelta(hours=self.last_hours or 6)
        if start.tzinfo is None:
            start = start.replace(tzinfo=dt.timezone.utc)
        if end.tzinfo is None:
            end = end.replace(tzinfo=dt.timezone.utc)
        if start >= end:
            raise ValidationError("window start must be before end")
        if (end - start) > dt.timedelta(hours=MAX_HOURS):
            raise ValidationError(f"window is limited to {MAX_HOURS} hours")
        return start, end


class Scope(BaseModel):
    """WHAT the widget is about — the no-code half of "no-code".

    Four ways to name a set of points, in narrowing order: an explicit list, one
    device, one category, or the whole estate. `category=""` is meaningful and
    selects the points nothing has classified, exactly as `/bi/devices` reads it.
    """

    model_config = ConfigDict(extra="forbid")

    type: Literal["points", "device", "category", "all"] = "points"
    point_ids: list[uuid.UUID] = Field(default_factory=list)
    device_id: uuid.UUID | None = None
    device_tag: str | None = None
    # "" selects the UNCLASSIFIED points; None means "no category filter".
    category: str | None = None

    def validate_for(self) -> None:
        if self.type == "points" and not self.point_ids:
            raise ValidationError("scope.type=points requires at least one point_id")
        if self.type == "device" and self.device_id is None and not self.device_tag:
            raise ValidationError("scope.type=device requires device_id or device_tag")
        if self.type == "category" and self.category is None:
            raise ValidationError("scope.type=category requires a category")


class WidgetQuery(BaseModel):
    """The executable half of the spec. Everything here reaches the database."""

    model_config = ConfigDict(extra="forbid")

    # The RESULT SHAPE, not the chart type. See the module docstring, rule 2.
    #   series    → buckets over time, one row per point (line, bar-over-time)
    #   aggregate → one value per group over the whole window (stat, bar, table)
    kind: Literal["series", "aggregate"] = "series"

    scope: Scope = Field(default_factory=Scope)

    # Which number. `count` is SAMPLES — the one quantity that is meaningful
    # across differently-measured points, which is why it is the only metric
    # `group_by` may widen past a single point.
    metric: Literal["avg", "min", "max", "last", "first", "count"] = "avg"

    # Which store answers. "auto" picks 1m up to three hours and 1h beyond,
    # exactly as /bi/series does, and reports which it used.
    rollup: Literal["auto", "1m", "1h", "raw"] = "auto"

    window: Window = Field(default_factory=Window)

    # `aggregate` only. point = one row per point; device / category roll up and
    # are restricted to metric="count" (see the module docstring).
    group_by: Literal["point", "device", "category"] = "point"

    limit: int = Field(default=12, ge=1, le=MAX_ROWS)

    def validated(self) -> "WidgetQuery":
        self.scope.validate_for()

        if self.kind == "series":
            if self.group_by != "point":
                raise ValidationError("a series widget always groups by point")
            if self.limit > MAX_SERIES:
                raise ValidationError(f"a series widget draws at most {MAX_SERIES} points")
            if len(self.scope.point_ids) > MAX_SERIES:
                raise ValidationError(f"a series widget draws at most {MAX_SERIES} points")

        if self.group_by in ("device", "category") and self.metric != "count":
            # The rule that keeps a bar chart from averaging a power factor with a
            # voltage. See the module docstring.
            raise ValidationError(
                f"metric '{self.metric}' cannot be grouped by {self.group_by}: values "
                "from different points are not comparable (no unit is on the wire). "
                "Use group_by=point, or metric=count to compare sample volume."
            )

        start, end = self.window.resolve()
        if self.rollup == "raw":
            span_min = (end - start).total_seconds() / 60.0
            if span_min > q.RAW_MAX_MINUTES:
                raise ValidationError(
                    f"raw readings are limited to {q.RAW_MAX_MINUTES} minutes "
                    f"(asked for {int(span_min)}); use rollup=1m or 1h"
                )
        return self


class WidgetSpec(BaseModel):
    """A whole widget definition, as stored on a dashboard row.

    `viz` is a plain string BY DESIGN — see the module docstring, rule 1. The
    frontend that ships with this build draws `line`, `bar`, `stat` and `table`;
    a build that adds `gauge` stores `viz="gauge"` and this executor is unchanged.
    """

    model_config = ConfigDict(extra="forbid")

    spec_version: int = SPEC_VERSION
    viz: str = "line"
    query: WidgetQuery = Field(default_factory=WidgetQuery)
    # Presentation only. Never reaches the database, never validated here — a
    # newer frontend's option key must not make an older backend reject the spec.
    options: dict = Field(default_factory=dict)


def _migrate(raw: dict) -> dict:
    """Bring a stored spec up to `SPEC_VERSION`.

    A no-op today because v1 is the only version that has ever existed. It is
    here, and called on every read, so that the FIRST time a field changes the
    upgrade lands in one place instead of being scattered through the executor —
    which is how saved dashboards quietly stop rendering.
    """
    version = raw.get("spec_version", SPEC_VERSION)
    if not isinstance(version, int):
        raise ValidationError("spec_version must be an integer")
    if version > SPEC_VERSION:
        raise ValidationError(
            f"this widget was saved by a newer version of the module "
            f"(spec_version {version}; this build understands {SPEC_VERSION})"
        )
    # v0 → v1 etc. would be applied here, in order.
    raw = dict(raw)
    raw["spec_version"] = SPEC_VERSION
    return raw


def _pydantic_message(exc: PydanticValidationError) -> str:
    """Turn pydantic's error list into one line naming the offending field.

    Without this the exception escapes as a 500: pydantic's ValidationError is
    only translated automatically when FastAPI itself validated the body, and
    this spec is parsed by hand (the route takes a raw dict so `_migrate` can run
    before validation). A widget editor needs to be told WHICH field it got
    wrong — "an unexpected error occurred" is unactionable.
    """
    parts = []
    for e in exc.errors()[:3]:
        loc = ".".join(str(x) for x in e.get("loc", ()) if x != "__root__")
        parts.append(f"{loc or 'spec'}: {e.get('msg', 'invalid')}")
    return "invalid widget spec — " + "; ".join(parts)


def parse(raw: dict) -> WidgetSpec:
    """Validate + migrate a stored or submitted spec. The ONE entry point."""
    if not isinstance(raw, dict):
        raise ValidationError("spec must be an object")
    try:
        spec = WidgetSpec.model_validate(_migrate(raw))
    except PydanticValidationError as exc:
        raise ValidationError(_pydantic_message(exc)) from exc
    spec.query = spec.query.validated()
    return spec


# ── Result shapes ────────────────────────────────────────────────────────────
#
# Two shapes, deliberately. Every widget type this module has (or gains) renders
# one of them, which is what decouples "new chart" from "backend change".


class ResultBucket(BaseModel):
    t: dt.datetime
    count: int
    min: float | None = None
    max: float | None = None
    avg: float | None = None
    first: float | None = None
    last: float | None = None
    txt_last: str | None = None


class ResultSeries(BaseModel):
    point_id: uuid.UUID
    point_tag: str | None
    device_tag: str | None
    # As STORED. NULL on this deployment and that is correct — never substituted.
    unit: str | None
    buckets: list[ResultBucket]


class ResultRow(BaseModel):
    """One aggregate row: a stat's number, a bar, a table line."""

    key: str
    label: str
    sublabel: str | None = None
    value: float | None = None
    # Samples behind `value` — how much the number is standing on. A stat tile
    # showing an average over three samples should be able to say so.
    samples: int = 0
    unit: str | None = None


class QueryResult(BaseModel):
    shape: Literal["series", "aggregate"]
    # Echoed so the renderer knows which bucket field to draw without re-parsing
    # the spec it just sent.
    metric: str
    resolution: str
    # One line of plain English the widget prints, so a chart never implies a
    # precision it does not have.
    resolution_reason: str
    start: dt.datetime
    end: dt.datetime
    # Exactly one of these is populated, per `shape`.
    series: list[ResultSeries] = Field(default_factory=list)
    rows: list[ResultRow] = Field(default_factory=list)
    # True when the scope matched more points than `limit` allowed — the widget
    # says "showing 12 of 37" instead of silently presenting a partial answer as
    # the whole one.
    truncated: bool = False
    matched: int = 0
